sympy/stats/joint_rv.py

```diff
@@ -304,13 +301,11 @@ def check(self):
 
     @property
     def set(self):
-        rvs = [i for i in random_symbols(self.args[1])]
-        marginalise_out = [i for i in random_symbols(self.args[1]) \
-         if i not in self.args[1]]
-        for i in rvs:
-            if i in marginalise_out:
-                rvs.remove(i)
-        return ProductSet((i.pspace.set for i in rvs))
+        rvs = set([i for i in random_symbols(self.args[1])])
```

Comment:
The order of elements in `rvs` is not canonical and this may be the source of the testing woes. Rather than modify the tests to hack around the indeterminacy, it is better to track down where it is coming from and fix it. In this case, perhaps, it is a matter of making `random_symbols` return a canonical list. I think it would be smart to do so since any code that relies on it does so without the benefit of seeing that it is returning unordered atoms.

Then the code here can be deconvoluted a bit:
1. set([i for i in random_symbols(foo)]) is just set(random_symbols(foo))
2. rvs is all the random variables in args[1], marginalise_out is all those that don't identically match something in args[1]; when you take these away from rvs you are left only with those items that *are* an identically matching rvs in args[1] so rvs can be computed once, I believe, as
```python
rvs = ordered(set(random_symbols(self.args[1])) & set(self.args[1]))
```
or more transparently as
```
rvs = [i for i in rvs if isinstance(i, RandomSymbol)]
```

Reply:
Ah! I see, I made mistake here, and therefore the results were randomised. Thanks a lot for helping.

Reply:
I think the problem is somewhere else. I tried out your changes and found out that the call to `symbols` or `set` is never made. And, one more point to be noted is that `marginalise_out` contains only one element for the tests in `test_mix.py`, so it is already ordered.

Reply:
> I think the problem is somewhere else

Did you try make the output of random_symbols canonical?

Reply:
Nice catch.

Reply:
I am a bit surprised that using `set` never caused a problem till now.